### PR TITLE
ENH: where for ufunc reductions

### PR DESCRIPTION
**Pull request by mhvk**
Note: This is a simpler and more robust, yet faster, version of #12635 and #12640 (way too much time wasted on being clever, but the comments helped steer to this better solution).

This introduces a `where` keyword for reductions. It works well, but for ufuncs with no identity, one has to explicitly pass in an `initial`. If people agree this is the way to go, I will add documentation and more tests.

Unlike for my other attempts, this performs quite well and I think we should consider using it in `nanfunctions` and in `MaskedArray`:
```
a = np.arange(100000.)
a[1000] = np.nan
np.add.reduce(a, where=~np.isnan(a)) == np.nansum(a)
# True
%timeit np.nansum(a)
# 10000 loops, best of 5: 113 µs per loop
%timeit np.add.reduce(a, where=~np.isnan(a))
# 10000 loops, best of 5: 92.3 µs per loop

# Now try what will be the worst-case scenario.
a[::2] = np.nan
np.add.reduce(a, where=~np.isnan(a)) == np.nansum(a)
# True
%timeit np.nansum(a)
# 1000 loops, best of 5: 381 µs per loop
%timeit np.add.reduce(a, where=~np.isnan(a))
# 1000 loops, best of 5: 389 µs per loop
```

### CODE REVIEW

File: numpy/core/src/umath/ufunc_object.c [hunk 1]
```diff
@@ -3531,8 +3534,49 @@ reduce_loop(NpyIter *iter, char **dataptrs, npy_intp *strides,
...
+
+        if (!masked) {
+            innerloop(dataptrs_copy, countptr,
+                      strides_copy, innerloopdata);
+        }
+        else {
+            npy_intp count = *countptr;
+            char *maskptr = dataptrs[2];
+            char mask = *maskptr;
+            if (strides[2] != 0) {
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Can you add a `mask_stride` alias for this?

File: numpy/core/src/umath/reduction.c [hunk 2]
```diff
@@ -524,8 +522,17 @@ PyUFunc_ReduceWrapper(PyArrayObject *operand, PyArrayObject *out,
...
     op_flags[1] = NPY_ITER_READONLY |
                   NPY_ITER_ALIGNED;
+    if (wheremask != NULL) {
+        op[2] = wheremask;
+        op_dtypes[2] = PyArray_DescrFromType(NPY_BOOL);
+        if (op_dtypes[2] == NULL) {
+            goto fail;
+        }
+        op_flags[2] = NPY_ITER_READONLY |
+                      NPY_ITER_ALIGNED;
```
Comment (eric-wieser) [inline numpy/core/src/umath/reduction.c]: What's the aligned for here?

File: numpy/core/src/umath/reduction.c [hunk 3]
```diff
@@ -493,6 +484,13 @@ PyUFunc_ReduceWrapper(PyArrayObject *operand, PyArrayObject *out,
         Py_INCREF(op_view);
     }
     else {
+        /* Cannot use where when we initialize from the operand */
```
Comment (eric-wieser) [inline numpy/core/src/umath/reduction.c]: This prevents passing `where=np.True_` into identity-less ufuncs, despite allowing `where=True`. I think we might want to detect empty masks on a per iterator-loop basis, to allow both.

Fine to leave to another PR.

File: numpy/core/src/umath/reduction.c [hunk 4]
```diff
@@ -493,6 +484,13 @@ PyUFunc_ReduceWrapper(PyArrayObject *operand, PyArrayObject *out,
...
+        if (wheremask != NULL) {
+            PyErr_SetString(PyExc_RuntimeError,
+                            "Reduce operations with no identity only support "
+                            "a where mask if 'initial' is passed in.");
```
Comment (eric-wieser) [inline numpy/core/src/umath/reduction.c]: Might be nice to include the ufunc name in this message.

File: numpy/core/tests/test_ufunc.py [hunk 5]
```diff
@@ -1758,7 +1778,7 @@ def test_reduce_arguments(self):
         # too little
         assert_raises(TypeError, f)
         # too much
-        assert_raises(TypeError, f, d, 0, None, None, False, 0, 1)
+        assert_raises(TypeError, f, d, 0, None, None, False, 0, True, 1)
```
Comment (eric-wieser) [inline numpy/core/tests/test_ufunc.py]: I think these arguments are swapped

**Review by eric-wieser — COMMENTED** [review]
Looks pretty good!

**Comment by njsmith** [comment]
Neat!

> for ufuncs with no identity, one has to explicitly pass in an `initial`. 

Can you explain why?

**Comment by njsmith** [comment]
Neat!

> for ufuncs with no identity, one has to explicitly pass in an `initial`. 

Can you explain why?

**Comment by seberg** [comment]
I think I may rather want to skip the `initial` part to leave the option of adding `identity` specifically.

**Comment by mhvk** [comment]
@shoyer - `initial` was initially called `identity` but after discussion was replaced only by `initial` - see #10635. In our current reductions, it is actually a better description of what happens...

@njsmith - the reason for having to pass in `initial` is the same as in my initial attempt, to be able to deal with the case where no elements are selected (but here, it is truly an initial, not an identity, as it is no longer used as a replacement).

File: numpy/core/src/umath/reduction.c (hunk 2 above)
Comment (mhvk) [reply numpy/core/src/umath/reduction.c]: Yes, I guess for bools that is pretty senseless. Removed.

File: numpy/core/src/umath/reduction.c (hunk 3 above)
Comment (mhvk) [reply numpy/core/src/umath/reduction.c]: Indeed, I prefer to leave this for another PR, as it needs some thought about what the ideal behaviour is (or even how to check whether an empty mask occurs).

**Comment by seberg** [comment]
@mhvk sorry, I think I interpreted it the wrong way around. I though `initial` was assumed to be an identity for the operation here.

File: numpy/core/tests/test_ufunc.py (hunk 5 above)
Comment (mhvk) [reply numpy/core/tests/test_ufunc.py]: It was for "too many arguments" which didn't work any more. But if we go with keyword only, I can revert this...

File: numpy/core/src/umath/reduction.c (hunk 4 above)
Comment (mhvk) [reply numpy/core/src/umath/reduction.c]: Yes. Actually, what error do you think this should be? `RunTimeError` is more for something that is unexpected, where this is not. Just `ValueError`? I guess I should catch it earlier, in the keyword getting stage (can keep it here just in case, though `ReduceWrapper` is not part of the C API so it is just for internal checking).

File: numpy/core/src/umath/ufunc_object.c (hunk 1 above)
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: Done

File: numpy/core/src/umath/ufunc_object.c [hunk 6]
```diff
@@ -4456,24 +4500,43 @@ PyUFunc_GenericReduction(PyUFuncObject *ufunc, PyObject *args,
...
+                                         &op,
+                                         &axes_in,
+                                         PyArray_DescrConverter2, &otype,
+                                         PyArray_OutputConverter, &out)) {
             goto fail;
         }
     }
     else {
-        if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|OO&O&iO:reduce",
+        if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|OO&O&iOO:reduce",
```
[the remark this answers is not in the document]
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: Yes, makes sense.

[the remark this answers is not in the document]
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: Actually, the way `sum` is done in `_methods.py`, this needs to be kept as a positional argument.

File: numpy/core/tests/test_ufunc.py (hunk 5 above)
Comment (eric-wieser) [reply numpy/core/tests/test_ufunc.py]: Oh, I see - the `1` is the extra. You're right, what you have is correct.

File: numpy/core/src/umath/ufunc_object.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_object.c]: Well, only if we want to keep the 10% performance boost. I suppose most users will pass by keyword-argument anyway, so there's little to gain by requiring it to be keyword-only.

**Comment by mhvk** [comment]
OK, now with initial comments addressed and more tests and docs.

**Comment by njsmith** [comment]
So if you have a reduction that has no identity, and you try to use `where=` but don't pass `initial=`, then is that always an error, or is it only an error when one of the reductions turns out to have zero elements?

**Comment by mhvk** [comment]
Right now, without an identity and initial, using `where` always errors. To change this would mean quite a bit of hacking in the code that determines the initial from the data itself: currently, it simply takes a view of the first element of the input data along the reduction axis, but obviously that won't work if that first element might be masked.  I would like to postpone this to later (or not do it at all; not sure it is worth it...).

**Comment by eric-wieser** [comment]
> not sure it is worth it

Without doing it, I don't think we can use it to implement things like summary masked arrays of dtype=object

File: numpy/core/_add_newdocs.py [hunk 7]
```diff
@@ -4876,7 +4876,7 @@
 
 add_newdoc('numpy.core', 'ufunc', ('reduce',
     """
-    reduce(a, axis=0, dtype=None, out=None, keepdims=False, initial)
+    reduce(a, axis=0, dtype=None, out=None, keepdims=False, initial, *, where=True)
```
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py]: This `*` is not true - I assume it's a remnant of the rejected suggestion to implement it this way.

File: numpy/core/src/umath/override.c [hunk 8]
```diff
@@ -257,8 +257,8 @@ normalize_reduce_args(PyUFuncObject *ufunc, PyObject *args,
             }
             obj = PyTuple_GetSlice(args, 3, 4);
         }
-        /* Remove initial=np._NoValue */
-        if (i == 5 && obj == NoValue) {
+        /* Remove {initial,where}=np._NoValue */
+        if (i >= 5 && obj == NoValue) {
```
Comment (eric-wieser) [inline numpy/core/src/umath/override.c]: Why is `where=np._NoValue` even allowed?

File: numpy/core/_add_newdocs.py (hunk 7 above)
Comment (mhvk) [reply numpy/core/_add_newdocs.py]: Oops, yes, corrected.

File: numpy/core/src/umath/override.c (hunk 8 above)
Comment (mhvk) [reply numpy/core/src/umath/override.c]: It gets used by `sum` etc in `fromnumeric`. In principle, though, those also strip it away again (unless overridden with `__array_function__`). Unlike `initial`, it does not get used in `_methods`, so I think it can indeed be removed. Will do so.

**Comment by mhvk** [comment]
> Without doing it, I don't think we can use it to implement things like summary masked arrays of dtype=object

I don't quite follow. Currently, the `MaskedArray.sum` just fills the masked pieces with `0`, so it is assumed that the object array knows what to do when summing with `0`. It would seem that nothing would break if one were to use `initial=0`.

p.s. For object arrays one could pass in a special `initial` object that has `def self.__add__(other): return other`, and then at the end replace any remaining instances of those with `masked`, thus avoiding the need to reduce the mask as well. But I don't see how to generalize this to non-object.

File: numpy/core/src/umath/ufunc_object.c [hunk 9]
```diff
@@ -3531,8 +3534,50 @@ reduce_loop(NpyIter *iter, char **dataptrs, npy_intp *strides,
...
+                    }
+                }
+                /* The number of items not yet processed */
+                count -= iprev;
+            }
+            /* If either the broadcast mask or the last one is false, we're done */
+            if (mask) {
+                innerloop(dataptrs_copy, &count,
+                          strides_copy, innerloopdata);
+            }
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Possibly clearer to transpose the two loops here:
```C++
npy_intp i = 0;
while (i < count) {
    /* Find where the mask changes */
    char mask = *maskptr;
    npy_intp i_next;
    if (mask_stride == 0) {
        /* optimization */
        i_next = count;
    }
    else {
        i_next = i;
        while (mask == *maskptr && i_next < count) {
            i_next++;
            maskptr += mask_stride;
        }
    }
    /* Copy the last contiguous region */
    npy_intp n = i_next - i;
    if (mask) {
        innerloop(dataptrs_copy, &n,
                  strides_copy, innerloopdata);
    }
    dataptrs_copy[0] += n * strides[0];
    dataptrs_copy[1] += n * strides[1];
    dataptrs_copy[2] = dataptrs_copy[0];
    i = i_next;
}
```

You can also leave the stride == 0 optimization to the compiler:
```C
while (count > 0) {
    /* Find where the mask changes */
    char mask = *maskptr;
    npy_intp n = find(maskptr, count, mask_stride, !mask);

    /* Copy the last contiguous region */
    if (mask) {
        innerloop(dataptrs_copy, &n,
                  strides_copy, innerloopdata);
    }
    dataptrs_copy[0] += n * strides[0];
    dataptrs_copy[1] += n * strides[1];
    maskptr          += n * mask_stride
    dataptrs_copy[2] = dataptrs_copy[0];
    count -= n;
}

static inline npy_intp
find(npy_bool *data, npy_intp n, npy_intp stride, bool value) {
    /* let the compiler optimize for zero stride */
    if (stride == 0){
        return find_inner(data, n, stride, value);
    }
    else {
        return find_inner(data, n, stride, value);
    }
}

static inline npy_intp
find_inner(npy_bool *data, npy_intp n, npy_intp stride, bool value) {
    for (npy_intp i = 0; i < n; i++, data += stride) {
        if (*data == value) {
            return i;
        }
    }
    return n;
}
```

Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: I like the version that does not rely on the compiler to optimize and adopted an even simpler version of it. It makes the case where the mask is all-true quite a bit faster:
```
a = np.arange(100000.)
m = ~np.isnan(a)
%timeit np.add.reduce(a, where=m)
# 90 -> 70 us
```
Adding either `a[::100] = np.nan` or `a[::2] = np.nan`, the timings stay basically the same.

Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_object.c]: To be clear - changing from the version this comment was added on to the latest version gives that speed boost? Or the `stride=0` optimization gives that speed up in either case?

**Comment by mhvk** [comment]
@eric-wieser - on `MaskedArray.sum`, I tried replacing `self.filled(0).sum(...)` with `self._data.sum(..., where=~_mask, initial=0)` and almost all masked array tests pass. The exception are tests with masked matrices, because I did not update `matrix.sum` to recognize `where` (or `initial` for that matter). Which I guess brings up a general problem of using this in `MaskedArray` - it would need to be able to fall back to using `filled` for a while to support other masked subclasses, while emitting a `FutureWarning` or so.

**Comment by mhvk** [comment]
p.s. I guess the first use of this may be in `nanfunctions`...

File: numpy/core/src/umath/ufunc_object.c (hunk 9 above)
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: The change in version (i.e., the last commit) is what gave the speed boost - it has nothing to do with the `stride=0` optimization (the mask has a stride in my example, it is just set to all True). I'm actually rather confused why there is such a speed-up, it must be that the compiler can make the inner `where` loop substantially faster than the large loop with a complex but unlikely if clause.

p.s. For the record, the timings for the `stride=0` case do not change.

Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_object.c]: > I'm actually rather confused why there is such a speed-up

So was I, hence the request for clarification!

**Comment by eric-wieser** [comment]
> For object arrays one could pass in a special initial object that has def self.__add__(other): return other

That's a really neat idea!

File: numpy/core/src/umath/ufunc_object.c [hunk 10]
```diff
@@ -4456,24 +4487,43 @@ PyUFunc_GenericReduction(PyUFuncObject *ufunc, PyObject *args,
...
+        if (where != NULL) {
+            PyArray_Descr *dtype;
+            dtype = PyArray_DescrFromType(NPY_BOOL);
+            if (dtype == NULL) {
+                goto fail;
+            }
+            /*
+             * Optimization: where=True is the same as no where argument.
+             * This lets us document it as a default argument.
+             */
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: This optimization needs to come before calling `PyArray_DescrFromType`, else you leak a `dtype`

Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_object.c]: Actually, I think you might leak it no matter what.

**Comment by eric-wieser** [comment]
Other than the reference leak above, this looks good to go for me

File: numpy/core/src/umath/ufunc_object.c (hunk 10 above)
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: `PyArray_FromAny` steals the reference to the `dtype`, so indeed the `dtype` should be created inside the `if` clause.

Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: Actually, turns out I wrote a `_wheremask_converter` when redoing the keyword parsing for the normal `ufunc`. Might as well use it...

**Comment by mhvk** [comment]
OK, fixed the reference issue by re-using `_wheremask_converter`, and rebased/squashed since it seems ready to go in.

**Comment by ahaldane** [comment]
Also, one case I thought about a lot has to do with non-commutative reductions. For instance:
```python
>>> np.equal.reduce([False, False], initial=False)                            
False
>>> np.equal.reduce([False, False, False], initial=False)
True
```
if you work it out, you'll see that the end result depends on how successive values alternate. Therefore, there is a difference between replacing a value by an identity element before reducing (my strategy), and not using the element in the reduction in the first place (this PR, correct?).

**Comment by eric-wieser** [comment]
> if you work it out, you'll see that the end result depends on how successive values alternat

Right, but `np.equal` has no identity. For the special case of the `??->?` loop, it has an identity of `True`, for which your above would work just fine.

**Comment by mhvk** [comment]
@ahaldane - yes, I think this should work well for the general masking case. The one problem I noted [earlier](https://github.com/numpy/numpy/pull/12644#issuecomment-451179117) is that whatever is being masked has to be able to deal with `where` and `initial` in their own `sum()` method. But if a project is all up to date with the numpy API, it should be fine.

p.s. Will look at your `MaskedArray` class; curious to see if it would allow astropy finally to have a `MaskedQuantity`...

**Review by eric-wieser — APPROVED** [review]

**Comment by mattip** [comment]
Nice and simple enhancement. Would be nice to add the comparison benchmark at some point

**Comment by seberg** [comment]
Yeah, looked good to me as well. Only thing to maybe quickly follow up on: Do we want to allow allow non-boolean masks? I think I would prefer to not allow it for now but the code seems to me like it does, and there should be a test for it probably.

**Comment by seberg** [comment]
Sorry, nevermind, the loop uses "safe" casting, thought I skimmed an unsafe cast there somewhere.

**Comment by mhvk** [comment]
> Nice and simple enhancement.  Would be nice to add the comparison benchmark at some point

I added a new issue #12662 to use this in `nanfunctions` - item 1 for that is to add benchmarks for those functions.

**Comment by mhvk** [comment]
@seberg - mask array creation and casting is now identical to what is used for the regular ufuncs, so OK to the extent those are OK!

**Comment by ahaldane** [comment]
I added a commit using it in the MaskedArray ducktype, and it passes my test suite and gives a fair speedup. Setup:
```python                                       
>>> np.random.seed(12345)                                                      
>>> x = np.random.rand(515, 512)                                                
>>> a = MaskedArray(x, x < 0.5) 
```

Before this PR:
```python
>>> %timeit np.add.reduce(a, axis=1)    
1.63 ms ± 8.96 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
After this PR:
```python                                      
>>> %timeit np.add.reduce(a, axis=1)                                           
1.39 ms ± 2.37 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

The case I tested might be a worst-case for this PR, too, since it has lots of alternating masked elements.

**Comment by mhvk** [comment]
@ahaldane - happy to see that it helps! And, yes, that'll definitely be close to worst-case performance.

**Comment by ahaldane** [comment]
Apologies, I mixed myself up on the timings. Here is the correct comparison:

Before:
```python
>>> a = MaskedArray(x, x < 0.9)                                                
>>> %timeit np.add.reduce(a, axis=1)                                           
751 µs ± 3.55 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
>>> a = MaskedArray(x, x < 0.5)
>>> %timeit np.add.reduce(a, axis=1)      
1.4 ms ± 3.07 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
>>> a = MaskedArray(x, x < 0.1)                                               
>>> %timeit np.add.reduce(a, axis=1)                                          
748 µs ± 16 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

After:
```python
>>> a = MaskedArray(x, x < 0.9)                                                
>>> %timeit np.add.reduce(a, axis=1)                                           
600 µs ± 1.97 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)    
>>> a = MaskedArray(x, x < 0.5)
>>> %timeit np.add.reduce(a, axis=1)      
1.63 ms ± 3.13 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)                       
>>> a = MaskedArray(x, x < 0.1)                                               
>>> %timeit np.add.reduce(a, axis=1)                                          
684 µs ± 15.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
So if I did the timings right it seems be faster in the two biased cases, but slower in the worst case.

**Comment by mhvk** [comment]
@ahaldane - OK, pity that it isn't always faster! I think in part it may be the final change of doing this in a nested `while` loop - the original I had of a single `for` loop turns out to have been faster for this case (but substantially worse for the presumably more common case where only few elements are not selected). Still, it means no need to make a copy, which should help for very large arrays (for `nansum`, things are faster again for `shape=(5120,5120)`; your case should be better since you only need to invert the mask, not also determine it using `isnan`).

**Comment by seberg** [comment]
Yeah, the only way to get better is probably to have specialized inner loop functions that handle the mask internally... Which makes it a ternary ufunc...

**Comment by mhvk** [comment]
@seberg - I've made some progress with "chaining ufuncs", which execute inner loops in sequence (see https://github.com/mhvk/chain_ufunc). I think those could solve problems like this too.

**Comment by seberg** [comment]
@mhvk yeah, chaining is another cool addition that is probably reasonably hard to achieve even. It could achieve something similar, although not for optimization purposes (that would basically need a dedicated specialization). I wonder how far we should go down the line doing things similar to numexpr or libraries that do lazy evaluation with optimization.

**Comment by mhvk** [comment]
Agreed that it is not clear how far one should go. The chaining I was looking at was mostly for use in quantities, where often you need to multiply one input with a constant to get the right units for the operation. It helps less than I hoped...
